консольное приложение №1

Nadezhda-Gurova opened the pull request:
Версия 1

Commit:
Add dev branch

Nadezhda-Gurova committed:
консольное приложение №1

SeveNChaK commented on src/main/java/Main.java:
```diff
@@ -1,8 +1,90 @@
+import java.util.Scanner;
+
 public class Main {
+    static class Product {
```
⚠️ Без особой необходимости не надо делать внутренние классы. Это усложняет понимание кода и может привести к нежелательным ошибкам. Лучше вынести его в отдельный файл, к тому же это просят в задании.

SeveNChaK commented on src/main/java/Main.java:
```diff
@@ -1,8 +1,90 @@
+    }
+
+
+    static void inputProduct(Scanner scanner, Basket basket) {
```
⏫ Сканер можно не передавать в качестве параметра везде, а просто вынести в поле класса `Main`

SeveNChaK commented on src/main/java/Main.java:
```diff
@@ -1,8 +1,90 @@
+            String name = scanner.nextLine();
+            System.out.println("Введите стоимость товара");
+            double price = Double.parseDouble(scanner.nextLine());
+            basket.add(new Product(name, price));
```
⚠️ Не хватает проверки на отрицательные числа
⚠️ Также не хватает проверки, если буду вводить буквы. Сейчас в такой ситуации приложение падает

SeveNChaK commented on src/main/java/Main.java:
```diff
@@ -1,8 +1,90 @@
+            char lastChar = rub.charAt(indexOf - 1);
+            if (lastChar == '1') {
+                return "рубль";
+            } else if ("234".indexOf(lastChar) != -1) {
```
⏫ Такой `indexOf` можно заменить на `contains`.
⚠️ Также для этих цифр склонение указано неверно - 2, 3, 4 рублЯ. И обрати внимание в какую ветку будут попадать такие числа, как 11-14 или 111-114. Кажется, что что-то будет не так. Стоит внимательнее обработать и проверить все варианты. Подсказка: надо бращать на последнюю или 2 последних цифры, в зависимости от ситуации

SeveNChaK commented on src/main/java/Main.java:
```diff
@@ -1,8 +1,90 @@
+
+        String inclineRubEnding(String rub) {
+            int indexOf = rub.indexOf(".");
+            char lastChar = rub.charAt(indexOf - 1);
```
⏫ Для нахождения нужных последних цифр лучше использовать остаток от деления. Например, `21 % 10` даст нам цифру 1, а `111 % 100` даст нам последние 2 цифры

SeveNChaK commented on src/main/java/Main.java:
```diff
@@ -1,8 +1,90 @@
+            try {
+                num = Integer.parseInt(scanner.nextLine());
+            } catch (NumberFormatException e) {
+
```
⏫ Неплохо бы что-нибудь тут написать. Например, сказать, что нужно вводить число, а не буквы

Nadezhda-Gurova committed:
- вынесла классы в отдельные файлы
- поправила ф-ию вычисления окончания
- проверила, что не вводят неккоректные значения

SeveNChaK commented on src/main/java/Basket.java:
```diff
@@ -0,0 +1,36 @@
+
+class Basket {
+    List<String> products = new ArrayList<>();
+    double sum = 0.0;
```
⏫ Все поля можно сделать приватными

SeveNChaK commented on src/main/java/Basket.java:
```diff
@@ -0,0 +1,36 @@
+    List<String> products = new ArrayList<>();
+    double sum = 0.0;
+
+    String inclineRubleEnding(int ruble) {
```
⏫ Можно сделать приватным

SeveNChaK commented on src/main/java/Basket.java:
```diff
@@ -0,0 +1,36 @@
+        for (String product : products) {
+            System.out.println(product);
+        }
+        System.out.println(String.format(Locale.US, "%.2f", sum / numberOfPersons)
```
⏫ Для информативности можно добавить текста. Например, "Каждый должен заплатить по:"

SeveNChaK commented on src/main/java/Basket.java:
```diff
@@ -0,0 +1,36 @@
+                + " " + inclineRubleEnding((int) Math.floor(sum / numberOfPersons)));
+    }
+
+    void add(Product product) {
```
⏫ В целом сюда можно просто сразу передать имя и цену товара и избавиться таким образов от класса `Product`. Просто больше этот класс нигде не используется, а создается и сразу же из него берутся поля, поэтому заворчивать их в этот класс - мало пользы
